Add e2e test for API coverage

roycaihw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Use dynamic client to simply test CRUD operations on all expected resources. Most of the YAML files are collected from kubernetes code base and [kubernetes.io](kubernetes.io). Currently the test covers 406 operations (test/e2e/testing-manifests/apiresource/resources.csv) out of 702 operations (test/e2e/testing-manifests/apiresource/resources_all.csv). 

The csv files are in the format of `group, version, resource, namespaced, verb`. The remaining resources are not tested yet because either the yaml file is missing, or the resource is non-namespaced and may break e2e test. Ref test/e2e/testing-manifests/apiresource/README.md for more information.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Ref kubernetes/test-infra#5446

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

/sig api-machinery
cc @erictune

roycaihw commented:
Hard-coded the file path on my local cluster but I guess the RepoRoot is deprecated/removed. Switching to go-bindata.

roycaihw commented:
/retest

roycaihw commented:
/cc @erictune @fejta @mithrav

fejta commented:
/uncc
/assign @lavalamp @deads2k 

Looks like apimachinery stuff?

fejta commented:
@k8s-mirror-api-machinery-pr-reviews

fejta commented:
@kubernetes/sig-api-machinery-pr-reviews

deads2k commented:
Wow, this looks like a maintenance nightmare.  What is is goal?  Is any/most of this generated?

@liggitt @derekwaynecarr called it?

thockin commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/57171)
<!-- Reviewable:end -->

bgrant0607 commented:
If we want to include this in conformance, one thing we'll need to figure out is how to exclude APIs that should not be covered. The existing coverage data already shows that some machinery is probing paths that should be excluded.

https://gist.github.com/oomichi/bc0044624443b4b596bdb4c83dd4ad4b

cc @kubernetes/sig-architecture-pr-reviews

roycaihw commented:
@deads2k The goal is to use a table driven test to verify if a Kubernetes cluster exposes endpoints an end user would reasonably expect. This test doesn't verify the correct behavior of those APIs, but just checks if the endpoints exist (as a necessary but not sufficient step). 

For the maintenance, may I ask if you are talking about the list of APIs or the yaml test data? The API list is generated by bringing up a cluster and using discovery API. The yaml test data is collected from the code base and kubernetes.io. It would be better if we could generate those yaml files from the schema/Go types. But if we manually maintain them, are we expecting the yaml file will be invalid (broken) for a beta/GA API frequently?

sttts commented on test/e2e/apimachinery/coverage.go:
```diff
@@ -0,0 +1,389 @@
+	subresources resourceMap
+}
+
+var _ = SIGDescribe("Coverage", func() {
```
This belong into some test/e2e/coverage folder. It tests coverage of Kubernetes APIs, it's not about generic API features.

sttts replied:
This comment still applies.

sttts commented:
> is generated by bringing up a cluster and using discovery API. The yaml test data is collected from the code base and kubernetes.io. It would be better if we could generate those yaml files from the schema/Go types. But if we manually maintain them, are we expecting the yaml file will be invalid (broken) for a beta/GA API frequently?

The question is who will own and update these files. API authors won't if there is no hack/verify-* mechanism which tells them so.

roycaihw commented:
> The question is who will own and update these files. API authors won't if there is no hack/verify-* mechanism which tells them so.

@sttts For API verbs that already exist in the coverage list, the test would fail if someone made some change in API that invalidated the yaml files. We could document the test and error messages to point to the API resource yaml file to let people fix it. For APIs that aren't under this coverage list yet, we need to establish some process to let API authors to add the API into the coverage list to expand the coverage. 

Since we are not testing the correct behavior but just the existence of APIs, we could focus on reviewing changes in the coverage list instead of changes under `test/e2e/testing-manifests/apiresource/yamlfiles`.

sttts commented:
> we need to establish some process to let API authors to add the API into the coverage list to expand the coverage.

This. We have to detect when something is missing, i.e. the coverage is not complete anymore. This is IMO a pre-requisite to move forward with this PR.

Just an idea: it might be easier to maintain if we declare what we explicitly do not want to have in conformance. If something new pops up (fields in types, endpoints) we can fail a PR if the coverage files are not updated. In other words, we would have two sets: 1) coverage 2) exceptions. And `1+2 = all exposed API` must holds and will be checked during CI.

roycaihw commented:
> We have to detect when something is missing, i.e. the coverage is not complete anymore

Agreed. We can have a script to discover all the endpoints by bringing up a local cluster, just like how we generate the openapi spec in `hack/update-openapi-spec.sh`, and verify the existing coverage list. I will update that in this PR. 

> If something new pops up (fields in types, endpoints) 

I suppose "fields in types" are about the yaml test data, not the coverage list. 

Also, the coverage list in this PR is not complete (see https://github.com/kubernetes/kubernetes/pull/57171/files#diff-399f824eb035a135055c8d83cffd9336) We could include these APIs in the exception list first.

roycaihw commented:
@sttts @deads2k I added a script to generate the list of all api resources served by apiserver, and added verification in CI. Feedback is welcome :)

`hack/update-api-resource-list.sh` and `hack/verify-api-resource-list.sh` generate and verify list of api resources served by kube apiserver.

`hack/update-api-resource-list.sh` is added to make rule `make verify_generated_files`, to get run in CI.

`test/e2e/testing-manifests/apiresource/resources_all.csv` is the generated list of all api resources served by kube apiserver.
`test/e2e/testing-manifests/apiresource/resources_whitelist.csv` is the whitelist of api verbs that we are not covering in api coverage test
`test/e2e/testing-manifests/apiresource/resources.csv` is the actual list of api verbs that we test in api coverage test, which is generated based on `resources_all.csv` and `resources_whitelist.csv`

/test pull-kubernetes-integration

roycaihw commented:
> This belong into some test/e2e/coverage folder. It tests coverage of Kubernetes APIs, it's not about generic API features.

@sttts I agree it's not about generic API features. Our [e2e test](https://github.com/kubernetes/community/blob/master/contributors/devel/e2e-tests.md#kinds-of-tests) document says every test should be owned by a SIG. This probably doesn't belong to sig-apimachinery. Maybe sig-architecture could own this? @bgrant0607

@deads2k @lavalamp Would like another round of review on this. Since the api resource list is auto-generated and verified in CI, we can expand the coverage by modifying the `resources_whitelist.csv` as future step.

sttts commented on build/root/Makefile.generated_files:
```diff
@@ -40,7 +40,8 @@ generated_files: gen_deepcopy gen_defaulter gen_conversion gen_openapi
                         verify_gen_defaulter   \
-                        verify_gen_conversion
+                        verify_gen_conversion  \
+			verify_gen_api_resource_list
```
is the indention right?

sttts commented on build/root/Makefile.generated_files:
```diff
@@ -659,6 +660,10 @@ gen_conversion: $(CONVERSION_FILES) $(CONVERSION_GEN)
 	$(RUN_GEN_CONVERSION) --verify-only
 
+.PHONY: verify_gen_api_resource_list
+verify_gen_api_resource_list:
```
is this verification only? I.e. there is no code generation involved? Then this probably belongs into hack/verify-all.sh

roycaihw replied:
Removed from generated_files make rules. The verification script is included in verify-all automatically (with verify- prefix)

sttts commented on hack/update-api-resource-list.sh:
```diff
@@ -0,0 +1,110 @@
+# Start kube-apiserver
+kube::log::status "Starting kube-apiserver"
+"${KUBE_OUTPUT_HOSTBIN}/kube-apiserver" \
+  --insecure-bind-address="${API_HOST}" \
```
this insecure bind will go away. I would prefer to not use it in new scripts.

roycaihw replied:
Updated to use TLS

sttts commented on hack/update-api-resource-list.sh:
```diff
@@ -0,0 +1,110 @@
+KUBE_RESOURCE_TEST_FILE="${KUBE_ROOT}/test/e2e/testing-manifests/apiresource/resources.csv"
+
+# Make sure subresource come after parent resource
+fetch_api_resource | sort -t',' -k1,1 -k2,2 -k3,3 -k4 -o ${KUBE_RESOURCE_FILE}
```
does this work with a BSD sort and grep on Mac?

roycaihw replied:
BSD sort and GNU sort give same result. BSD grep doesn't work well with -x flag. Switched to use `comm` which works consistently in both GNU and BSD

sttts commented on hack/verify-api-resource-list.sh:
```diff
@@ -0,0 +1,68 @@
+
+mkdir -p "${_tmp}"
+cp -a "${LISTROOT}" "${TMP_LISTROOT}"
+trap "cp -a ${TMP_LISTROOT} ${LISTROOT}/..; rm -rf ${_tmp}" EXIT SIGINT
```
don't we have a general trap mechanism? This command overrides any trap installed before.

roycaihw replied:
Updated to use util function to add trap

sttts commented on test/e2e/apimachinery/coverage.go:
```diff
@@ -0,0 +1,389 @@
+	if group == "" {
+		group = "core"
+	}
+	yamlFile := generated.ReadOrDie(filepath.Join("test/e2e/testing-manifests/apiresource/yamlfiles/", fmt.Sprintf("%s/%s/%s", group, r.version, r.name)+".yaml"))
```
this breaks on Windows with `/`, I would expect

roycaihw replied:
This "filename" is used as gobindata index, which gets converted into slash-style path in the end https://github.com/kubernetes/kubernetes/blob/e6b6e5c4b4f501daa61ce0554b36931104565cab/vendor/github.com/jteeuwen/go-bindata/toc.go#L165-L167

Added `filepath.FromSlash()` to make it more readable for Windows

roycaihw commented:
@sttts Thanks for the feedback! I've addressed the comments, PTAL :)

sttts commented:
/assign @cblecker 

for another bash review.

roycaihw commented:
@mml for bash review

mml commented on hack/update-api-resource-list.sh:
```diff
@@ -0,0 +1,114 @@
+KUBE_APISERVER_CERT="$TMP_DIR/certs/apiserver.crt"
+AUTH_TOKEN_HEADER="Authorization: Bearer dummy_token"
+
+function fetch_api_resource()
```
Document the format this returns?

mml commented on hack/update-api-resource-list.sh:
```diff
@@ -0,0 +1,114 @@
+sort -t',' -k1,1 -k2,2 -k3,3 -k4 ${KUBE_RESOURCE_WHITELIST_FILE} -o ${KUBE_RESOURCE_WHITELIST_FILE}
+# GNU comm and BSD comm have different defaulting. We sort the inputs to satisfy 
+# the defaulting first, then sort the output to more readable csv format
+comm -13 <(sort ${KUBE_RESOURCE_WHITELIST_FILE}) <(sort ${KUBE_RESOURCE_FILE}) | sort -t',' -k1,1 -k2,2 -k3,3 -k4 -o ${KUBE_RESOURCE_TEST_FILE}
```
Add a comment here explaining the intent, which is to remove whitelist lines from RESOURCE_FILE

mml commented on hack/update-api-resource-list.sh:
```diff
@@ -0,0 +1,114 @@
+KUBE_RESOURCE_WHITELIST_FILE="${KUBE_ROOT}/test/e2e/testing-manifests/apiresource/resources_whitelist.csv"
+KUBE_RESOURCE_TEST_FILE="${KUBE_ROOT}/test/e2e/testing-manifests/apiresource/resources.csv"
+
+# Make sure subresource come after parent resource
```
I don't understand this comment.

mml commented on hack/verify-api-resource-list.sh:
```diff
@@ -0,0 +1,70 @@
+# limitations under the License.
+
+# Script to fetch latest api resources served by kube apiserver
+# Puts the updated list at test/e2e/testing-manifests/apiresource/resources_all.csv
```
Comment is wrong for this file.

roycaihw commented:
@mml Thanks for the comments. Updated

mml commented:
/lgtm

roycaihw committed:
Add scripts to generate list of api resources

`hack/update-api-resource-list.sh` and
`hack/verify-api-resource-list.sh`
generate and verify list of api resources served by kube apiserver.

`test/e2e/testing-manifests/apiresource/resources_all.csv` is the
generated list of all api resources served by kube apiserver.

`test/e2e/testing-manifests/apiresource/resources_whitelist.csv` is the
whitelist of api verbs that we are not covering in api coverage test

`test/e2e/testing-manifests/apiresource/resources.csv` is the actual
list of api verbs that we test in api coverage test, which is generated
based on `resources_all.csv` and `resources_whitelist.csv`

roycaihw committed:
Prepare API resources yaml files using bindata

roycaihw committed:
Add API coverage test

roycaihw commented:
@mml Squashed. PTAL

/test pull-kubernetes-e2e-kops-aws

mml commented:
/lgtm

roycaihw commented:
/assign @sttts @cblecker 
for approval

roycaihw committed:
Rebase and update api resource list

roycaihw committed:
Documentation

roycaihw commented:
From a test infra perspective, the value from verifying existence of API CRUD operations doesn't match the effort of hand-crafting object yaml files that exercise API validation. Nevertheless, we should think about having better documentation & examples in our OpenAPI models and verbs. Ref https://github.com/kubernetes/kubernetes/issues/34743, https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/

From conformance perspective, we should focus on testing the correct API behavior on pluggable parts of the system, and increasing coverage with those tests. Ref https://github.com/kubernetes/test-infra/issues/5446#issuecomment-393753951

Closing
